File: solutions/[105]从前序与中序遍历序列构造二叉树.cpp
#include <vector>
#include <map>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
		map<int, int> m;
		for (int i = 0; i < inorder.size(); ++i)
			m[inorder[i]] = i;
		int pre_idx = 0;
		return dp(m, preorder, inorder, pre_idx, 0, inorder.size());
	}
	TreeNode* dp(map<int, int>& m, vector<int>& preorder, vector<int>& inorder, int& pre_idx, int in_left, int in_right) {
		if (in_left == in_right)
			return nullptr;

		TreeNode* root = new TreeNode(preorder[pre_idx]);
		int idx = m[preorder[pre_idx]];
		pre_idx++;
		root->left = dp(m, preorder, inorder, pre_idx, in_left, idx);
		root->right = dp(m, preorder, inorder, pre_idx, idx + 1, in_right);

		return root;
	}
};

File: solutions/[105]从前序与中序遍历序列构造二叉树.py
# Definition for a binary tree node.
from typing import List


# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        val2p_in = dict()
        for i, val in enumerate(inorder):
            val2p_in[val] = i
        def build_tree(p_pre, l_in, r_in):
            if l_in > r_in:
                return None
            root_val = preorder[p_pre]

            p_root_in = val2p_in[root_val]
            root = TreeNode(root_val,
                            build_tree(p_pre+1, l_in, p_root_in-1),
                            build_tree(p_root_in-l_in+p_pre+1, p_root_in+1, r_in))
            return root
        
        return build_tree(0, 0, len(inorder)-1)
